### PR TITLE
[SPARK-47692][SQL] Fix default StringType meaning in implicit casting

### PR DESCRIPTION
**Pull request by mihailomilosevic2001**
### What changes were proposed in this pull request?
Addition of priority flag to StringType.

### Why are the changes needed?
In order to follow casting rules for collations, we need to know whether StringType is considered default, implicit or explicit.

### Does this PR introduce _any_ user-facing change?
Yes.

### How was this patch tested?
Implicit tests in CollationSuite.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by mihailom-db** [comment]
This PR merged in changes from https://github.com/apache/spark/pull/45383 in order to build features faster.

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala [hunk 1]
```diff
@@ -27,7 +34,10 @@ import org.apache.spark.sql.catalyst.util.CollationFactory
  * @param collationId The id of collation for this StringType.
  */
 @Stable
-class StringType private(val collationId: Int) extends AtomicType with Serializable {
+class StringType private(val collationId: Int,
+                         var priority: StringTypePriority = ImplicitST)
+  extends AtomicType
```
Comment (HyukjinKwon) [inline sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala:39]:
```suggestion
class StringType private(
    val collationId: Int,
    var priority: StringTypePriority = ImplicitST)
  extends AtomicType
```

File: sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala [hunk 2]
```diff
@@ -76,8 +86,10 @@ class StringType private(val collationId: Int) extends AtomicType with Serializa
  * @since 1.3.0
  */
 @Stable
-case object StringType extends StringType(0) {
-  private[spark] def apply(collationId: Int): StringType = new StringType(collationId)
+case object StringType extends StringType(0, ImplicitST) {
+  private[spark] def apply(collationId: Int,
+                           priority: StringTypePriority = ImplicitST): StringType =
```
Comment (HyukjinKwon) [inline sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala:91]:
```suggestion
  private[spark] def apply(
      collationId: Int, priority: StringTypePriority = ImplicitST): StringType =
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  override val transform: PartialFunction[Expression, Expression] = {
+    case e if !e.childrenResolved => e
+    case sc@(_: In
+             | _: InSubquery
+             | _: CreateArray
+             | _: ComplexTypeMergingExpression
+             | _: ArrayJoin
+             | _: BinaryExpression
+             | _: ConcatWs
+             | _: Substring) =>
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala:139]:
```suggestion
    case sc @ (
        _: In | _: InSubquery | _: CreateArray | _: ComplexTypeMergingExpression |
        _: ArrayJoin | _: BinaryExpression | _: ConcatWs | _: Substring) =>
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 4]
```diff
@@ -608,9 +609,11 @@ abstract class TypeCoercionBase {
       case c @ Concat(children) if !c.childrenResolved || children.isEmpty => c
       case c @ Concat(children) if conf.concatBinaryAsString ||
         !children.map(_.dataType).forall(_ == BinaryType) =>
+        val st = getOutputCollation(c.children)
         val newChildren = c.children.map { e =>
-          implicitCast(e, StringType).getOrElse(e)
+          implicitCast(e, st).getOrElse(e)
         }
+
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:616]:
```suggestion
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 5]
```diff
@@ -733,7 +737,6 @@ abstract class TypeCoercionBase {
               udfInputToCastType(in.dataType, expected.asInstanceOf[DataType])
             ).getOrElse(in)
           }
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:739]:
```suggestion
          }

```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala [hunk 6]
```diff
@@ -20,421 +20,406 @@ package org.apache.spark.sql
...
+    with SharedSparkSession
+    with ExpressionEvalHelper {
 
   case class CollationTestCase[R](s1: String, s2: String, collation: String, expectedResult: R)
   case class CollationTestFail[R](s1: String, s2: String, collation: String)
 
   test("Support Like string expression with Collation") {
+    def prepareLike(input: String,
+                    regExp: String,
+                    collation: String): Expression = {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala:41]:
```suggestion
    def prepareLike(
        input: String,
        regExp: String,
        collation: String): Expression = {
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala [hunk 7]
```diff
@@ -20,421 +20,406 @@ package org.apache.spark.sql
...
         )
       )
-    })
+    )
   }
 
   test("Support ILike string expression with Collation") {
+    def prepareILike(input: String,
+                    regExp: String,
+                    collation: String): Expression = {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala:78]:
```suggestion
    def prepareILike(
        input: String,
        regExp: String,
        collation: String): Expression = {
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala [hunk 8]
```diff
@@ -20,421 +20,406 @@ package org.apache.spark.sql
...
         )
       )
-    })
+    )
   }
 
   test("Support RLike string expression with Collation") {
+    def prepareRLike(input: String,
+                    regExp: String,
+                    collation: String): Expression = {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala:117]:
```suggestion
    def prepareRLike(
        input: String,
        regExp: String,
        collation: String): Expression = {
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala [hunk 9]
```diff
@@ -20,421 +20,406 @@ package org.apache.spark.sql
...
         )
       )
-    })
+    )
   }
 
   test("Support StringSplit string expression with Collation") {
+    def prepareStringSplit(input: String,
+                           splitBy: String,
+                           collation: String): Expression = {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala:155]:
```suggestion
    def prepareStringSplit(
        input: String,
        splitBy: String,
        collation: String): Expression = {
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala [hunk 10]
```diff
@@ -20,421 +20,406 @@ package org.apache.spark.sql
...
         )
       )
-    })
+    )
   }
 
   test("Support RegExpExtract string expression with Collation") {
+    def prepareRegExpExtract(input: String,
+                             regExp: String,
+                             collation: String): RegExpExtract = {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala:233]:
```suggestion
    def prepareRegExpExtract(
        input: String,
        regExp: String,
        collation: String): RegExpExtract = {
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala [hunk 11]
```diff
@@ -20,421 +20,406 @@ package org.apache.spark.sql
...
         )
       )
-    })
+    )
   }
 
   test("Support RegExpExtractAll string expression with Collation") {
+    def prepareRegExpExtractAll(input: String,
+                             regExp: String,
+                             collation: String): RegExpExtractAll = {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala:272]:
```suggestion
    def prepareRegExpExtractAll(
        input: String,
        regExp: String,
        collation: String): RegExpExtractAll = {
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala [hunk 12]
```diff
@@ -20,421 +20,406 @@ package org.apache.spark.sql
...
         )
       )
-    })
+    )
   }
 
   test("Support RegExpCount string expression with Collation") {
+    def prepareRegExpCount(input: String,
+                           regExp: String,
+                           collation: String): Expression = {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala:311]:
```suggestion
    def prepareRegExpCount(
        input: String,
        regExp: String,
        collation: String): Expression = {
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala [hunk 13]
```diff
@@ -20,421 +20,406 @@ package org.apache.spark.sql
...
         )
       )
-    })
+    )
   }
 
   test("Support RegExpSubStr string expression with Collation") {
+    def prepareRegExpSubStr(input: String,
+                           regExp: String,
+                           collation: String): Expression = {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala:350]:
```suggestion
    def prepareRegExpSubStr(
        input: String,
        regExp: String,
        collation: String): Expression = {
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala [hunk 14]
```diff
@@ -20,421 +20,406 @@ package org.apache.spark.sql
...
         )
       )
-    })
+    )
   }
 
   test("Support RegExpInStr string expression with Collation") {
+    def prepareRegExpInStr(input: String,
+                                regExp: String,
+                                collation: String): RegExpInStr = {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationRegexpExpressionsSuite.scala:389]:
```suggestion
    def prepareRegExpInStr(
        input: String,
        regExp: String,
        collation: String): RegExpInStr = {
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 15]
```diff
@@ -20,57 +20,60 @@ package org.apache.spark.sql
...
 
-class CollationStringExpressionsSuite extends QueryTest
-  with SharedSparkSession with ExpressionEvalHelper {
+class CollationStringExpressionsSuite
+  extends QueryTest
+    with SharedSparkSession
+    with ExpressionEvalHelper {
 
   case class CollationTestCase[R](s1: String, s2: String, collation: String, expectedResult: R)
+
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:36]: can you remove unrelated changes?

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 16]
```diff
@@ -20,57 +20,60 @@ package org.apache.spark.sql
...
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.test.SharedSparkSession
 import org.apache.spark.sql.types.StringType
 
-class CollationStringExpressionsSuite extends QueryTest
-  with SharedSparkSession with ExpressionEvalHelper {
+class CollationStringExpressionsSuite
+  extends QueryTest
+    with SharedSparkSession
+    with ExpressionEvalHelper {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:33]:
```suggestion
  with SharedSparkSession
  with ExpressionEvalHelper {
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 17]
```diff
@@ -20,57 +20,60 @@ package org.apache.spark.sql
...
   case class CollationTestCase[R](s1: String, s2: String, collation: String, expectedResult: R)
+
   case class CollationTestFail[R](s1: String, s2: String, collation: String)
 
+
   test("Support ConcatWs string expression with Collation") {
-    // Supported collations
+    def prepareConcatWs(sep: String,
+                        collation: String,
+                        inputs: Any*): ConcatWs = {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:43]:
```suggestion
    def prepareConcatWs(
        sep: String,
        collation: String,
        inputs: Any*): ConcatWs = {
```

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 18]
```diff
@@ -509,6 +497,209 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
...
+        s"WHERE utf8_binary_lcase IN " +
+        s"('aaa' COLLATE UTF8_BINARY_LCASE, 'bbb' COLLATE UTF8_BINARY_LCASE)"),
+        Seq(Row("aaa", "aaa"), Row("AAA", "AAA"), Row("bbb", "bbb"), Row("BBB", "BBB")))
+      checkAnswer(sql(s"SELECT * FROM $tableName " +
+        s"WHERE utf8_binary_lcase IN ('aaa' COLLATE UTF8_BINARY_LCASE, 'bbb')"),
+        Seq(Row("aaa", "aaa"), Row("AAA", "AAA"), Row("bbb", "bbb"), Row("BBB", "BBB")))
+    }
+  }
+
+  // TODO: (SPARK-47210) Add indeterminate support
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:675]:
```suggestion
  // TODO(SPARK-47210): Add indeterminate support
```

**Comment by HyukjinKwon** [comment]
Let's follow https://github.com/databricks/scala-style-guide, and remove unrelated changes, e.g., adding newlines which makes cherry-pick/backporting/reverting difficult.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 19]
```diff
@@ -645,6 +646,34 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
...
+      withSQLConf(SqlApiConf.DEFAULT_COLLATION -> "UNICODE") {
+        checkAnswer(spark.sql("SELECT collation(:var1 || :var2)",
+          Map(
+            "var1" -> Literal.create("a", StringType(1)),
+            "var2" -> Literal.create("b", StringType(2))
+          )
+        ),
+          Seq(Row("UNICODE"))
+        )
+      }
```
Comment (mihailom-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:676]: @srielau Is this the expected behaviour? Apparently we can pass different collations to parameters. I understood that the behaviour should be if StringType has priority of default then it has to have session level default collation, as otherwise we might have different collations with same default priority, which is not covered by design.

Comment (mihailom-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:676]: Why I ask this is because by implementation we are using Parameter which binds an Expression to a name/postition depending on the query type. Allowed expression can be CreateArray, which produces implicit collated elementType, should we cast it to default if the value comes from a parameter. The thing is that if I only cast the flag to defaultST, I might get two arrays of different collationId, but same priority of DafaultST. Current implementation casts any StringType or ArrayType(StringType) to session level default StringType/ ArrayType with that session level default StringType.

**Comment by mihailom-db** [comment]
@dbatomic @stefankandic @uros-db @stevomitric @nikolamand-db  Could you take a look at this PR? It introduces a new flag for collations and makes sure most of the functions that return StringType without StringType in inputs return default collation.

**Comment by nikolamand-db** [comment]
How do we want to handle cases like string [split](https://github.com/apache/spark/pull/45856/files#diff-8ce4728fc96a1a5f2a3a846470f03175ba00b61ff6bc51933883241e736d886aR548) (work in progress, used as illustration) in general? The regex parameter accepts only `StringType` and if I understand the plan would fail when we explicitly set the default collation type to be something other than `UTF8_BINARY` and run query like `select split('abc', 'b')`. However, in this case the regex parameter's collation doesn't have any semantic meaning and I don't think the queries alike should fail.

**Comment by mihailom-db** [comment]
If the collation is not important to the function execution we can create a separate pattern for it where we will exclude parameters that do not need casting. For example, substring is done in that way. We only cast if it is DefaultST or ExplicitST to ImplicitST, as we need the return type to have implicit collation.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala (hunk 19 above)
Comment (srielau) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:676]: I don't know your internal design (and I shouldn't need to), so I won't comment on that.
But these examples seem correct. What is this 1/2 business in StringType() about?
Basically the user has two choices:
* say nothing: parameter marker goes with DEFAULT and looses against columns
* explicitly collate: parameter marker goes with explicit collation and wins against columns.

And if you re-use the same marker in multiple places you can choose differently.

**Review by uros-db — APPROVED** [review]
lgtm

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala [hunk 20]
```diff
@@ -118,17 +121,21 @@ object CollationTypeCasts extends TypeCoercionRule {
           )
       // Only implicit or default collations present
       case 0 =>
-        val implicitTypes = expr.map(_.dataType)
+        val implicitTypes = expr.filter {
+            case Literal(_, _: StringType) => false
+            case Cast(child, _: StringType, _, _) => child.dataType.isInstanceOf[StringType]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala:126]: Let's also check `cast.getTagValue(Cast.USER_SPECIFIED_CAST).isEmpty` to make sure it's not a user-specified cast.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala:126]: oh wait, if users use scala/python DataFrame to cast a column to a StringType with specific collation, should it be explicit collation?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala:126]: Same for SQL: `CAST(col AS STRING WITH ...)`

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
